---
title: "workflowR"
author: "Julia Arukakkal"
date: "2024-06-18"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

# workflowR {.tabset}
## workflowR About  
Hi everyone

## Steps of TOGA

1. First, TOGA uses pairwise genome alignment between the reference & query genome represented by chains of co-linear local alignments. Through this process, it captures orthologous gene loci and loci containing paralogs or processed pseudo-genes. To distinguish between these 2 groups, TOGA computes characteristic features that capture the amount of intronic and intergenic alignments, considering each gene & overlapping chain. Additionally, synteny (conserved gene order) can be used to help further distinguish orthologs from paralogs. TOGA also implements machine learning to compute the probability that a chain is an orthologous locus for the gene in question. 

  - The most important feature for classification is the capture of    intronic and intergenic alignments, leaving synteny as the least    important. Hence, synteny is used as an auxiliary feature that      allows TOGA to detect orthologs that underwent translocation or     inversion.

  * <details>
     <summary>More on the machine learning feature</summary>
       
       The machine learning classifier was trained on known orthologous genes between humans (reference) and mice (query) from Ensembl Compara. This classifier was then tested on independent query species (rat, dog, and armadillo) that represented different placental mammal orders  and produced a nearly perfect classification of orthologous chains. 
       
       When manually examining data it was found that misclassifications that produced false positives are mostly due to partial or full gene duplications and half of false negative may be due to faster X chromosome evolution. 
       
   </details>
   
2. CESAR v. 2.0 is used to determine the position of coding exons of the focal gene in each (co-)orthologous query locus. TOGA then assesses the reading frame intactness for each transcript and applies a gene loss detection approach to identify gene-inactivating mutations while taking assembly incompleteness into account. A gene is classified as *lost* if all transcripts at all (co-)orthologous loci are classified as lost.

  * <details>
      <summary>More on CESAR v. 2.0</summary>
      "CESAR 2.0 (Coding Exon Structure Aware Realigner 2.0) is a          method to realign coding exons or genes to DNA sequences            using a Hidden Markov Model.  

      Compared to its predecessor, CESAR 2.0 is 77X times faster       on average (132X times faster for large exons) and requires         30-times less memory. In addition, CESAR 2.0 improves the           accuracy of the comparative gene annotation by two new              features. First, CESAR 2.0 substantially improves the               identification of splice sites that have shifted over a larger       distance, which improves the accuracy of detecting the correct       exon boundaries. Second, CESAR 2.0 provides a new gene mode         that re-aligns entire genes at once. This mode is able to           recognize complete intron deletions and will annotate larger        joined exons that arose by intron deletion events."
      
      -- From the [Hiller Lab GitHub page on CESAR 2.0](https://github.com/hillerlab/CESAR2.0)
      
    </details>
    
3. In this step, TOGA determines the orthology type by considering all reference genes and orthologous query loci that encode within an intact reading frame. TOGA then uses an orthology graph to resolve any weakly supported orthology relationship among many:many orthologs.


```{r}

```

